### PR TITLE
[Streaming]Streaming queue support failover

### PR DESCRIPTION
**Pull request by MissiontoMars**
## Why are these changes needed?
This PR add failover support in streaming queue, preparing for the following failover feature of data transfer.

Specifically, in this PR, downstream queues can pull data from upstream queues after recovery from failover.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by MissiontoMars** [comment]
@chaokunyang @ashione

File: streaming/src/channel.cc [hunk 1]
```diff
@@ -148,24 +150,44 @@ StreamingQueueConsumer::~StreamingQueueConsumer() {
   STREAMING_LOG(INFO) << "Consumer Destroy";
 }
 
-StreamingStatus StreamingQueueConsumer::CreateTransferChannel() {
+StreamingQueueStatus StreamingQueueConsumer::GetQueue(
+    const ObjectID &queue_id, uint64_t start_msg_id,
+    ChannelCreationParameter &init_param) {
```
Comment (ashione) [inline streaming/src/channel.cc:155]: const ChannelCreationParameter &init_param

File: streaming/src/channel.cc [hunk 2]
```diff
@@ -148,24 +150,44 @@ StreamingQueueConsumer::~StreamingQueueConsumer() {
...
+  STREAMING_LOG(INFO) << "GetQueue qid: " << queue_id << " start_msg_id: " << start_msg_id
+                      << " actor_id: " << init_param.actor_id;
   auto downstream_handler = ray::streaming::DownstreamQueueMessageHandler::GetService();
-  STREAMING_LOG(INFO) << "GetQueue qid: " << channel_info_.channel_id
-                      << " start_seq_id: " << channel_info_.current_seq_id + 1;
-  if (downstream_handler->DownstreamQueueExists(channel_info_.channel_id)) {
-    RAY_LOG(INFO) << "StreamingQueueReader::GetQueue duplicate!!!";
-    return StreamingStatus::OK;
+  if (downstream_handler->DownstreamQueueExists(queue_id)) {
+    RAY_LOG(INFO) << "StreamingQueueReader:: Already got this queue.";
```
Comment (ashione) [inline streaming/src/channel.cc:160]: use same logger  STREAMING_LOG or RAY_LOG

File: streaming/src/protobuf/streaming_queue.proto [hunk 3]
```diff
@@ -67,4 +73,34 @@ message StreamingQueueTestInitMsg {
 message StreamingQueueTestCheckStatusRspMsg {
   string test_name = 1;
   bool status = 2;
+}
+
+message StreamingQueuePullRequestMsg {
```
Comment (ashione) [inline streaming/src/protobuf/streaming_queue.proto:78]: redundant definition of src_actor_id, dst_actor_id, queue_id, msg_id.
How about use an unified proto?

File: streaming/src/queue/message.h [hunk 4]
```diff
@@ -135,15 +147,100 @@ class CheckRspMessage : public Message {
...
+  PullResponseMessage(const ActorID &actor_id, const ActorID &peer_actor_id,
+                      const ObjectID &queue_id, uint64_t seq_id, uint64_t msg_id,
+                      queue::protobuf::StreamingQueueError err_code,
+                      bool is_upstream_first_pull)
+      : Message(actor_id, peer_actor_id, queue_id),
+        seq_id_(seq_id),
+        msg_id_(msg_id),
+        is_upstream_first_pull_(is_upstream_first_pull),
+        err_code_(err_code) {}
+  virtual ~PullResponseMessage() {}
```
Comment (ashione) [inline streaming/src/queue/message.h:188]: suggestion : 
```
virtual ~PullResponseMessage() = defult;
```

File: streaming/src/queue/queue.cc [hunk 5]
```diff
@@ -169,6 +168,115 @@ void WriterQueue::OnNotify(std::shared_ptr<NotificationMessage> notify_msg) {
...
+int WriterQueue::ResendItems(std::list<QueueItem>::iterator start_iter,
+                             uint64_t first_seq_id, uint64_t last_seq_id) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  int count = 0;
+  auto it = start_iter;
+  for (; it != watershed_iter_; it++) {
+    if (it->SeqId() > last_seq_id) {
+      break;
+    }
+    STREAMING_LOG(INFO) << "ResendItems send seq_id " << (*it).SeqId() << " to peer.";
```
Comment (ashione) [inline streaming/src/queue/queue.cc:191]: (*it).SeqId() -> it->Seq_id()

File: streaming/src/queue/queue.cc [hunk 6]
```diff
@@ -169,6 +168,115 @@ void WriterQueue::OnNotify(std::shared_ptr<NotificationMessage> notify_msg) {
...
+    count++;
+  }
+
+  STREAMING_LOG(INFO) << "ResendItems total count: " << count;
+  is_resending_ = false;
+  return count;
+}
+
+void WriterQueue::FindItem(
+    uint64_t target_msg_id, std::function<void()> large_callback, std::function<void()> small_callback,
```
Comment (ashione) [inline streaming/src/queue/queue.cc:202]: It's confused to use callback like `large_callback` and `small_callback`.

File: streaming/src/queue/queue_handler.cc [hunk 7]
```diff
@@ -108,6 +117,21 @@ void QueueMessageHandler::Stop() {
...
 
+void UpstreamQueueMessageHandler::Start() {
+  STREAMING_LOG(INFO) << "UpstreamQueueMessageHandler::Start";
+  QueueMessageHandler::Start();
+  handle_service_thread_ = std::thread([this] { handler_service_.run(); });
+}
+
+void UpstreamQueueMessageHandler::Stop() {
+  STREAMING_LOG(INFO) << "UpstreamQueueMessageHandler::Stop";
+  handler_service_.stop();
```
Comment (ashione) [inline streaming/src/queue/queue_handler.cc:128]: check handle_service_thread_ and make sure it's non-null.

File: streaming/src/channel.cc (hunk 1 above)
Comment (MissiontoMars) [reply streaming/src/channel.cc]: fixed.

File: streaming/src/protobuf/streaming_queue.proto (hunk 3 above)
Comment (MissiontoMars) [reply streaming/src/protobuf/streaming_queue.proto:76]: `MessageCommon` added.

File: streaming/src/queue/queue.cc (hunk 6 above)
Comment (MissiontoMars) [reply streaming/src/queue/queue.cc]: Better to use `greater` and `less`.

File: streaming/src/queue/queue.cc [hunk 8]
```diff
@@ -111,27 +109,28 @@ size_t Queue::PendingCount() {
...
 
-  while (is_pulling_) {
-    STREAMING_LOG(INFO) << "This queue is sending pull data, wait.";
+  while (is_resending_) {
+    STREAMING_LOG(INFO) << "This queue is resending data, wait.";
     std::this_thread::sleep_for(std::chrono::milliseconds(10));
   }
 
-  QueueItem item(seq_id, data, data_size, timestamp, raw);
+  QueueItem item(seq_id, buffer, buffer_size, timestamp, msg_id_start, msg_id_end, raw);
```
Comment (lixin-wei) [inline streaming/src/queue/queue.cc:123]: `seq_id_` instead of `seq_id `

Comment (lixin-wei) [inline streaming/src/queue/queue.cc:123]: `seq_id_` instead of `seq_id`

**Review by lixin-wei — DISMISSED** [review]

File: streaming/src/channel.cc (hunk 2 above)
Comment (MissiontoMars) [reply streaming/src/channel.cc]: Fixed.

File: streaming/src/queue/queue.cc (hunk 5 above)
Comment (MissiontoMars) [reply streaming/src/queue/queue.cc]: Fixed

File: streaming/src/queue/queue.cc (hunk 8 above)
Comment (MissiontoMars) [reply streaming/src/queue/queue.cc:125]: Fixed

File: streaming/src/channel.cc [hunk 9]
```diff
@@ -57,7 +57,7 @@ StreamingStatus StreamingQueueProducer::CreateQueue() {
                       << " data_size: " << channel_info_.queue_size;
   auto upstream_handler = ray::streaming::UpstreamQueueMessageHandler::GetService();
   if (upstream_handler->UpstreamQueueExists(channel_info_.channel_id)) {
-    RAY_LOG(INFO) << "StreamingQueueWriter::CreateQueue duplicate!!!";
+    STREAMING_LOG(INFO) << "StreamingQueueWriter::CreateQueue duplicate!!!";
```
Comment (ashione) [inline streaming/src/channel.cc:60]: StreamingQueueWriter::CreateQueue duplicated

File: streaming/src/queue/queue.h [hunk 10]
```diff
@@ -61,32 +67,37 @@ class Queue {
   /// Return the last item in pending state.
   QueueItem BackPending();
 
-  bool IsPendingEmpty();
-  bool IsPendingFull(uint64_t data_size = 0);
+  inline bool IsPendingEmpty();
```
Comment (ashione) [inline streaming/src/queue/queue.h:70]: Should be full implementation for inline function

File: streaming/src/queue/message.h [hunk 11]
```diff
@@ -55,24 +56,36 @@ class Message {
...
+        msg_id_start_(msg_id_start),
+        msg_id_end_(msg_id_end),raw_(raw) {}
   virtual ~DataMessage() {}
 
   static std::shared_ptr<DataMessage> FromBytes(uint8_t *bytes);
   virtual void ToProtobuf(std::string *output);
-  uint64_t SeqId() { return seq_id_; }
-  bool IsRaw() { return raw_; }
-  queue::protobuf::StreamingQueueMessageType Type() { return type_; }
+  inline uint64_t SeqId() { return seq_id_; }
```
Comment (chaokunyang) [inline streaming/src/queue/message.h:79]: Do we still have seq id?

**Review by chaokunyang — APPROVED** [review]
LGTM

File: streaming/src/queue/message.h (hunk 11 above)
Comment (MissiontoMars) [reply streaming/src/queue/message.h:79]: Yes.

**Review by raulchen — APPROVED** [review]

**Comment by rkooo567** [comment]
Hi, Hao. This PR seemed to break lint. Please make sure all necessary builds are passing before merging the PR!

**Pull request merged**